core/src/main/scala/com/ubirch/avatar/core/kafka/KafkaConsumer.scala

```diff
@@ -0,0 +1,143 @@
+    val messageInfo =
+      s"topic: ${message.record.topic}, group: $groupName, partition: ${message.committableOffset.partitionOffset.key}, offset: ${message.committableOffset.partitionOffset.offset}"
+
+    def retry(currentDelay: Int, retriesLeft: Int): Future[Unit] = {
+      handleMessage(message.record.value())
+        .map { _ =>
+          logger.info(s"succeeded to process message. $messageInfo")
+        }
+        .recoverWith {
+          case InvalidDataException(err) =>
+            logger.error(s"Invalid data exception occurred. Won't be retried. $err. $messageInfo")
+            Future.successful(())
+          case err =>
+            if (retriesLeft > 0) {
+              val nextDelay = Math.ceil(currentDelay * backoffFactor).toInt.min(maxBackoff)
+              logger.warn(
+                s"An unexpected exception occurred because of ${err.getMessage}, retry after ${currentDelay.millisecond}[ms]($retriesLeft retries left). $messageInfo")
+              after(currentDelay.millisecond, s)(retry(nextDelay, retriesLeft - 1))
+            } else {
+              logger.error(
+                s"Giving up after $maxRetries attempts because of ${err.getMessage}. this message will be committed. message: $message. $messageInfo")
+              Future.successful(())
+            }
+        }
+    }
+
+    retry(initialBackoff, maxRetries)
+  }
+
+  /**
+    * this is a supervision strategy for commit error.
+    *
+    * @Reference: https://github.com/akka/alpakka-kafka/issues/750
+    */
+  private val resumeOnCommitFailed: Supervision.Decider = {
+    // Supervision.Resume means that skips commit of current message and process next message.
+    case _: CommitFailedException =>
+      logger.error("commit failed exception occurred. skip the commit of the current message. actor will resume.")
+      Supervision.Resume
+    case ex =>
+      logger.error(s"unexpected error occurred. actor will stop. error $ex", ex)
+      Supervision.Stop
+  }
+
+  /**
+    * This method makes consumer running
+    */
+  def runWithRetry(kafkaRetryConfig: KafkaRetryConfig)(
+    implicit executionContext: ExecutionContext): DrainingControl[Done] = {
+    logger.info(s"$topicNames consumer starts running.")
+    kafkaSource()
+      .log(name = s"$topicNames consumer")
+      .addAttributes(
+        Attributes.logLevels(
+          onElement = Attributes.LogLevels.Off,
+          onFinish = Attributes.LogLevels.Off,
+          onFailure = Logging.ErrorLevel
+        ))
+      .mapAsync(Config.kafkaSubscribeParallel) { msg =>
+        val result =
+            retryHandler(
+              msg,
+              kafkaRetryConfig.maxRetries,
+              kafkaRetryConfig.backoffFactor,
+              kafkaRetryConfig.minBackoff,
+              kafkaRetryConfig.maxBackoff)
+        result.map { _ =>
+          msg.committableOffset
+        }
+      }
+      .toMat(Committer.sink(committerSettings.withMaxBatch(Config.kafkaMaxCommit))
```

Comment:
in version alpakka kafka 1, it doesn't retry when fails to commit even if the error is retriable.
https://github.com/akka/alpakka-kafka/pull/1111 

So if we use still version 1, it's better to catch `RetriableCommitFailedException` exception and commit again manually.